### PR TITLE
Added MORP text classification example, modified from Flax SST2 example.

### PR DESCRIPTION
**Pull request by jasonbaldridge**
Simple text classification example to get modeling for Rendezvous started!

This script trains a text classifier on a dataset containing paragraphs that
are linked to either Manhattan or Pittburgh. This is just to get something
basic that works for the geo-oriented text problems we are addressing with
the Rendezvous task.

A sentence is encoded with an LSTM and a binary prediction is made from
the final LSTM state using an MLP.

The data is loaded from raw text files and converted into TF Records.

Example command line call:

```
$ bazel-bin/cabby/model/text/train \
  --data_dir ~/data/morp/morp-balanced  \
  --model_dir ~/tmp/morp-model
```

This code is copied with very few modifications from the Flax SST2 example:

https://github.com/google/flax/tree/master/examples/sst2

The main thing was to load from text files instead of TFDS data resource. Model code and text processing all remain unchanged.

### CODE REVIEW

File: cabby/model/text/input_pipeline.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+      },
+      drop_remainder=False).prefetch(tf.data.experimental.AUTOTUNE)
+
+def _get_label_index(label):
+  """Assign index 0 to pittsburgh and 1 to manhattan."""
+  return tf.cast(label == 'manhattan', tf.int32)
+
+def _parse_line_as_example(line):
+  """Parses line with 'label<TAB>paragraph' into separate label and text."""
+  tabsep_strings = tf.strings.split(line, maxsplit=1)
```
Comment (sayalikulkarni) [inline cabby/model/text/input_pipeline.py:116]: do we have tabs in the data?

File: cabby/model/text/model.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+  module = TextClassifier.partial(train=False, **model_kwargs)
+  _, initial_params = module.init_by_shape(
+      jax.random.PRNGKey(seed),
+      [((batch_size, max_len), jnp.int32),
+       ((batch_size,), jnp.int32)])
+  model = nn.Model(module, initial_params)
+  return model
+
+
+def word_dropout(inputs: jnp.ndarray, rate: float, unk_idx: int,
```
Comment (sayalikulkarni) [inline cabby/model/text/model.py:44]: is there a default value for rate? may be move it as the second last arg?

File: cabby/model/text/model.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+  return model
+
+
+def word_dropout(inputs: jnp.ndarray, rate: float, unk_idx: int, 
+        deterministic: bool = False):
+  """Replaces a fraction (rate) of inputs with <unk>."""
+  if deterministic or rate == 0.:
+    return inputs
+
+  mask = jax.random.bernoulli(nn.make_rng(), p=rate, shape=inputs.shape)
```
Comment (sayalikulkarni) [inline cabby/model/text/model.py:50]: not sure - but is this costly? can we init it once as a constant and then permute for every example?

File: cabby/model/text/model.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+  return jnp.where(mask, jnp.array([unk_idx]), inputs)
```
Comment (sayalikulkarni) [inline cabby/model/text/model.py:51]: mask token should ideally be different than the unk.

File: cabby/model/text/train.py [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+flags.DEFINE_integer(
+    'batch_size', default=64,
+    help=('Batch size for training.'))
+
+flags.DEFINE_integer(
+    'num_epochs', default=20,
+    help=('Number of training epochs.'))
+
+flags.DEFINE_float(
+    'dropout', default=0.5,
```
Comment (sayalikulkarni) [inline cabby/model/text/train.py:77]: wow this looks like a very high value for dropout! may be better to have a lower default.

File: cabby/model/text/train.py [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+    help=('Dropout rate'))
+
+flags.DEFINE_float(
+    'emb_dropout', default=0.5,
```
Comment (sayalikulkarni) [inline cabby/model/text/train.py:81]: same here

File: cabby/model/text/train.py [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+flags.DEFINE_integer(
+    'hidden_size', default=256,
+    help=('Hidden size for the LSTM and MLP.'))
+
+flags.DEFINE_integer(
+    'embedding_size', default=256,
+    help=('Size of the word embeddings.'))
+
+flags.DEFINE_integer(
+    'max_seq_len', default=55,
```
Comment (sayalikulkarni) [inline cabby/model/text/train.py:105]: 55 is based on?
mean lengths in our dataset is about ~30, median ~27

File: cabby/model/text/train.py [hunk 8]
```diff
@@ -0,0 +1,396 @@
...
+flags.DEFINE_float(
+    'l2_reg', default=1e-6,
+    help=('L2 regularization weight'))
+
+flags.DEFINE_integer(
+    'seed', default=0,
+    help=('Random seed for network initialization.'))
+
+flags.DEFINE_integer(
+    'checkpoints_to_keep', default=1,
```
Comment (sayalikulkarni) [inline cabby/model/text/train.py:121]: prefer to have 3 or 5 as default value.

File: cabby/model/text/train.py [hunk 9]
```diff
@@ -0,0 +1,396 @@
...
+    help=('How many checkpoints to keep. Default: 1 (keep best model only)'))
+
+
+@jax.vmap
+def binary_cross_entropy_loss(logit: jnp.ndarray, label: jnp.ndarray):
```
Comment (sayalikulkarni) [inline cabby/model/text/train.py:126]: not a review comment -- but jax does not have a default impl for xent loss? :(

**Review by sayalikulkarni — CHANGES_REQUESTED** [review]
Looks nice! some minor inputs/questions

File: cabby/model/text/train.py (hunk 9 above)
Comment (jasonbaldridge) [reply cabby/model/text/train.py:126]: Jax certainly doesn't as it is more "bare metal" that neural network libraries build abstraction upon. Flax appears not to have default implementations for loss functions, but Trax does: 

https://trax-ml.readthedocs.io/en/latest/trax.layers.html#module-trax.layers.metrics

Not sure if they are planning to add loss functions to Flax, but do know they are working on the new Linen API for Flax: https://github.com/google/flax/tree/master/flax/linen

**Review by jasonbaldridge — COMMENTED** [review]
Thanks for the comments! I agree with you that the defaults could be better, but I want to keep things as close as possible to the Flax SST2 example so that we can update as/when that changes:

https://github.com/google/flax/tree/master/examples/sst2

(E.g. a minor change was made just yesterday!)

I commented on the data format. LMK if there is anything else you need for approval!

**Review by sayalikulkarni — APPROVED** [review]
Thanks Jason.
Re: defaults -- sgtm!
Re: unk_idx for mlm - fine to leave as is for now. but worth considering that this can also be used for regularization - so would be good to distinguish the 2. iiuc <unk> is not for reg. <mask> would be.
